contracts/modules/cover/Cover.sol

```diff
@@ -186,343 +185,185 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard {
-    );
-
-    // i think this require redundant now because the staking pool would have reverted
-    require(totalCoveredAmountInCoverAsset >= params.amount, "Cover: The selected pools ran out of capacity");
-
-    _coverSegments[coverId].push(CoverSegment(
-        totalCoveredAmountInCoverAsset, // amount
-        uint32(block.timestamp + 1), // start
-        SafeUintCast.toUint32(params.period), // period
-        _productTypes[_products[params.productId].productType].gracePeriodInDays,
-        priceRatio,
-        false, // expired,
-        globalRewardsRatio
-      ));
-
-    return (totalPremiumInNXM, totalCoveredAmountInCoverAsset);
-  }
-
-  function allocateCapacity(
-    BuyCoverParams memory params,
-    uint coverId,
-    IStakingPool _stakingPool,
-    uint amount
-  ) internal returns (uint coveredAmountInNXM, uint premiumInNXM, uint rewardsInNXM) {
-
-    Product memory product = _products[params.productId];
-    uint gracePeriod = uint(_productTypes[product.productType].gracePeriodInDays) * 1 days;
-
-    return _stakingPool.allocateStake(
-      CoverRequest(
-        coverId,
-        params.productId,
-        amount,
-        params.period,
-        gracePeriod,
-        globalCapacityRatio,
-        product.capacityReductionRatio,
-        globalRewardsRatio
-      )
-    );
-  }
-
-  function editCover(
-    uint coverId,
-    BuyCoverParams memory buyCoverParams,
-    PoolAllocationRequest[] memory poolAllocations
-  ) external payable onlyMember nonReentrant whenNotPaused {
-
-    CoverData memory cover = _coverData[coverId];
-    uint lastCoverSegmentIndex = _coverSegments[coverId].length - 1;
-    CoverSegment memory lastCoverSegment = coverSegments(coverId, lastCoverSegmentIndex);
-
-    require(ICoverNFT(coverNFT).isApprovedOrOwner(msg.sender, coverId), "Cover: Only owner or approved can edit");
-    require(buyCoverParams.period < MAX_COVER_PERIOD, "Cover: Cover period is too long");
-    require(buyCoverParams.commissionRatio <= MAX_COMMISSION_RATIO, "Cover: Commission rate is too high");
-
-    // Override cover specific parameters
-    buyCoverParams.coverAsset = cover.coverAsset;
-    buyCoverParams.productId = cover.productId;
-
-    Product memory product = _products[buyCoverParams.productId];
-    require(!product.isDeprecated, "Cover: Product is deprecated");
-
-    uint32 deprecatedCoverAssetsBitmap = pool().deprecatedCoverAssetsBitmap();
-
-    // Check that the payout asset is still supported (it may have been deprecated or disabled in the meantime)
-    require(
-      isAssetSupported(_getSupportedCoverAssets(deprecatedCoverAssetsBitmap, product.coverAssets), buyCoverParams.coverAsset),
-      "Cover: Payout asset is not supported"
-    );
-    require(
-      !_isCoverAssetDeprecated(deprecatedCoverAssetsBitmap, buyCoverParams.paymentAsset),
-      "Cover: payment asset deprecated"
-    );
+    {
+      require(_products.length > params.productId, "Cover: Product not found");
```

Comment:
What's the reason for not having all these validations at he beggining to reject the tx faster if something is wrong?

Reply:
Moved up